### ISSUE TITLE
KDE conflicts with stylix resulting in inconsistent theme

### ISSUE THREAD

**Issue by mrnetlex** [danth/stylix#267]
At every reboot something (I would suspect `kde-gtk-config`) overwrites `.gtkrc-2.0` and my cursor, fonts revert to those set in system settings. I can improve situation by removing `.gtkrc-2.0` and running `nixos-rebuild switch`, but then cursor theme apply only to GTK programs, but not any QT programs and  plasma elements like.

My configuration:
```nix
{
  #Stylix
  stylix.image = /home/netlex/NixOS/wallpapers/Tokyonight/cafe-at-night_00_3840x2160.png; 
  stylix.polarity = "dark"; 
  stylix.base16Scheme = "${pkgs.base16-schemes}/share/themes/tokyodark-terminal.yaml"; 
  stylix.cursor = {
    package = pkgs.qogir-icon-theme;
    name = "Qogir";
  };
  stylix.fonts = {
    serif = {
      package = pkgs.nerdfonts;
      name = "FireCode Nerd Font";
    };

    sansSerif = {
      package = pkgs.nerdfonts;
      name = "Ubuntu Nerd Font";
    };

    monospace = {
      package = pkgs.nerdfonts;
      name = "JetBrainsMono Nerd Font";
    };

    emoji = {
      package = pkgs.noto-fonts-emoji;
      name = "Noto Color Emoji";
    };
  };

  stylix = {
      opacity = let
        opacity_value = 0.8;
      in
        {
          applications = opacity_value;
          desktop = opacity_value;
          popups = opacity_value;
          terminal = opacity_value;
        };
    };

}
```

**Comment by trueNAHO** [comment]
As a side note,

```nix
stylix.image = /home/netlex/NixOS/wallpapers/Tokyonight/cafe-at-night_00_3840x2160.png;
```

might be [considered as impure by Flakes](https://github.com/tweag/rfcs/blob/ca2845b81814c345945de4be0b36cd5fb4eb080b/rfcs/0049-flakes.md?plain=1#L417-L420).

**Comment by LovingMelody** [comment]
Relative paths are still impure, wallpaper just needs to be git tracked or fetched from an online source

**Comment by Kitt3120** [comment]
Same problem here. I installed stylix into my NixOS configuration using the **non-flake** variant. Except this issue, everything works.

Every time I do a `nixos-rebuild switch`, and stylix applies its targets because something like my wallpaper changed, it works **only once**. After that, the `.gtkrc-2.0` file in my home exists. Because of this, the home manager service for my user can't come back up, cause it tries to also write into that file, but the existing file is in its way.

For example, this does not just happen after a reboot, but also when changing wallpaper and applying my config again with `nixos-rebuild switch`. If I don't manually delete the .gtkrc-2.0 in my home folder manually, the home manager service fails because `.gtkrc-2.0` is already there. This still happens even when setting `stylix.targets.gtk.enable = false`, so maybe something with the Kde Plasma module? Or maybe the files are being generated by Plasma itself.

**Comment by danth** [comment]
>  Instead, there is now a GNOME/GTK theme sync background service running with Plasma 6.

This is the same as with Plasma 5; that's what the `kded5rc` setting is turning off. I can confirm that it worked in Plasma 5, at least after manually removing the file the first time (since the service isn't disabled until after applying Stylix).

The KDE module hasn't been updated at all for Plasma 6 yet.

**Comment by Kitt3120** [comment]
> >  Instead, there is now a GNOME/GTK theme sync background service running with Plasma 6.
> 
> This is the same as with Plasma 5; that's what the `kded5rc` setting is turning off. I can confirm that it worked in Plasma 5, at least after manually removing the file the first time (since the service isn't disabled until after applying Stylix).
> 
> The KDE module hasn't been updated at all for Plasma 6 yet.

Ohh alright. I forked the project and imported Stylix from my local checkout, but I just couldn't get it to generate the kded6rc file. Maybe I was doing something wrong. I also noticed that the testbed file does not have the updated attributes yet. Do you want me to make a PR tomorrow with my current state and you can take it from there?

**Comment by danth** [comment]
I don't have much time to work on this myself right now, but feel free to open one anyways.

> I also noticed that the testbed file does not have the updated attributes yet.

If you post this as a separate PR, that part should be OK to merge straight away.

**Comment by mogery** [comment]
> It can be disabled by going into `System Settings -> Session (very bottom category) -> Background Services -> GNOME/GTK Settings Synchronization Service`. With it disabled, stylix works fine now. Maybe someone finds a solution to disabled this through the stylix KDE module? If not, this should at least be mentioned somewhere in the docs or in the README.

I *think* you can disable this via [plasma-manager](https://github.com/pjones/plasma-manager). I'll update this comment when trying it at home today.

**Comment by Kitt3120** [comment]
Made a PR for this, turns out testbeds work way different than I thought. The PR can't be merged this way, but I made some suggestions in that PR, so I would appreciate checking it out.

**Comment by hasrthur** [comment]
> > It can be disabled by going into `System Settings -> Session (very bottom category) -> Background Services -> GNOME/GTK Settings Synchronization Service`. With it disabled, stylix works fine now. Maybe someone finds a solution to disabled this through the stylix KDE module? If not, this should at least be mentioned somewhere in the docs or in the README.
> 
> I _think_ you can disable this via [plasma-manager](https://github.com/pjones/plasma-manager). I'll update this comment when trying it at home today.

@mogery hi! have you managed to disable that with plasma manager?

**Comment by hasrthur** [comment]
> > > It can be disabled by going into `System Settings -> Session (very bottom category) -> Background Services -> GNOME/GTK Settings Synchronization Service`. With it disabled, stylix works fine now. Maybe someone finds a solution to disabled this through the stylix KDE module? If not, this should at least be mentioned somewhere in the docs or in the README.
> > 
> > 
> > I _think_ you can disable this via [plasma-manager](https://github.com/pjones/plasma-manager). I'll update this comment when trying it at home today.
> 
> @mogery hi! have you managed to disable that with plasma manager?

I came up with such a solution:

```nix
programs.plasma.configFile.kded5rc = {
   "Module-gtkconfig"."autoload" = false;
};
```

**Comment by ccalhoun1999** [comment]
Any solutions to this without plasma-manager right now?
This option no longer seems to exist (at least I can't find it):
`System Settings -> Session (very bottom category) -> Background Services -> GNOME/GTK Settings Synchronization Service`

**Comment by eliasp** [comment]
The corresponding KCM isn't [shown in System Settings anymore](https://invent.kde.org/plasma/systemsettings/-/merge_requests/333) in Plasma 6.
It can be launched standalone now instead (Runner → `Background Services`, alternatively `kcmshell6 kcm_kded`).

**Comment by jmickelin** [comment]
I don't trust the `Module-*` autoload settings in `kded*rc`, because `kded` [sometimes chooses to completely ignore it](https://github.com/Alexays/Waybar/issues/3468#issuecomment-4408904301) (section "Why is the module not being loaded?").

So I tried to come up with a couple of alternative solutions (without requiring `plasma-manager`).

# Solution 1: Making the module write to a different file (rough idea)
Looking at the source code for `kde-gtk-config`, you should be able to trick it into barfing out its config file in a different location (which you could then `include` in your normal `.gtkrc-2.0`/`.config/gtk-2.0/gtkrc` file, should you want to). 

When writing the `gtk-2.0` RC file, it [reads the location from the `GTK2_RC_FILES` environment variable](https://github.com/KDE/kde-gtk-config/blob/de1a2b72f99b71cfd8839f6168be9c2446976aa8/kded/config_editor/gtk2.cpp#L40). (Though, do note that it ignores it entirely if the environment variable it contains multiple paths).

This means that you can wrap the `kded6` binary using [`wrapProgram`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/setup-hooks/make-wrapper.sh) with `--set GTK2_RC_FILES /some/other/path/gtkrc-2.0`, and it should write to that path instead. (I _think_ the path needs to be writable, because `kde-gtk-config` is a bit of a mess and fails to check return values where it should.)

Then, optionally, add a line like `include "/some/other/path/gtkrc-2.0"` to the Nix-generated `gtkrc` file, to load whatever the KDE-generated settings are.

## Limitations
I discovered it uses the same pattern that removes symlinks and replaces them with its own files for pretty much all the other settings it implements syncing for. This includes:
- `.config/xsettingsd/xsettingsd.conf`
- `.config/gtk-$VERSION/settings.ini`
- Asset files `.config/gtk-$VERSION/assets/*`
- CSS files `.config/gtk-$VERSION/{gtk.css,colors.css,window_decorations.css}`

for all `$VERSION`s 2.0, 3.0, 4.0.

I think the only reason that it _always_ modifies the GTK 2.0 RC file when the daemon is started (as opposed to waiting until you change any settings in the KDE menus) is because it performs some migrations to remove deprecated options. Though it might do the same stuff for the later versions, but for those it uses `gsettings` instead of files, which isn't as apparent. I haven't checked to make sure.

# Solution 2: Patching the module
I wasn't happy with the above limitations, so I wrote a patch that first makes it check whether a file it wants to modify is a symlink pointing to the Nix store. This means it will continue to work for files that aren't handled by your Nix config, but if you ever decide to Nixify any of them, it won't fight you anymore:

**kde-gtk-config-respect-store-symlinks.patch**:

<details>

```patch
diff --git a/kded/config_editor/custom_css.cpp b/kded/config_editor/custom_css.cpp
index eea00e3..986d588 100644
--- a/kded/config_editor/custom_css.cpp
+++ b/kded/config_editor/custom_css.cpp
@@ -28,12 +28,15 @@ void saveWindowDecorationsToAssets(const QStringList &windowDecorationsButtonsIm
 
         for (const auto &buttonImagePath : windowDecorationsButtonsImages) {
             const QString destination = assetsFolder.path() + '/' + QFileInfo(buttonImagePath).fileName();
-            QFile(destination).remove();
-            QFile(buttonImagePath).copy(buttonImagePath, destination);
+            QFile file(destination);
+            if (Utils::removeUnlessNixStorePath(file)) {
+              QFile(buttonImagePath).copy(buttonImagePath, destination);
+            };
         }
 
         for (const auto &buttonImagePath : windowDecorationsButtonsImages) {
-            QFile(buttonImagePath).remove();
+            QFile file(buttonImagePath);
+            Utils::removeUnlessNixStorePath(file);
         }
     }
 }
@@ -45,8 +48,10 @@ void addWindowDecorationsCssFile()
 
         auto windowDecorationsDestination = Utils::configDirPath(gtkVersion) + QStringLiteral("/window_decorations.css");
 
-        QFile(windowDecorationsDestination).remove();
-        windowDecorationsCss.copy(windowDecorationsDestination);
+        QFile file(windowDecorationsDestination);
+        if (Utils::removeUnlessNixStorePath(file)) {
+          windowDecorationsCss.copy(windowDecorationsDestination);
+        };
     }
 }
 
@@ -69,9 +74,10 @@ void addImportStatementsToGtkCssUserFile()
                 }
             }
 
-            gtkCss.remove();
-            gtkCss.open(QIODevice::WriteOnly | QIODevice::Text);
-            gtkCss.write(gtkCssContents);
+            if (Utils::removeUnlessNixStorePath(gtkCss)) {
+              gtkCss.open(QIODevice::WriteOnly | QIODevice::Text);
+              gtkCss.write(gtkCssContents);
+            };
         }
     }
 }
@@ -80,7 +86,8 @@ void removeWindowDecorationsCSS()
 {
     for (auto gtkVersion : Utils::s_gtkVersions) {
         QFile windowsDecorationsCss(Utils::configDirPath(gtkVersion) + QStringLiteral("/window_decorations.css"));
-        windowsDecorationsCss.remove();
+
+        Utils::removeUnlessNixStorePath(windowsDecorationsCss);
     }
 }
 
diff --git a/kded/config_editor/gtk2.cpp b/kded/config_editor/gtk2.cpp
index aad8209..9120ede 100644
--- a/kded/config_editor/gtk2.cpp
+++ b/kded/config_editor/gtk2.cpp
@@ -44,9 +44,11 @@ void setValue(const QString &paramName, const QVariant &paramValue)
     QFile gtkrc(gtkrcPath);
     QString gtkrcContents = Utils::readFileContents(gtkrc);
     replaceValueInGtkrcContents(gtkrcContents, paramName, paramValue);
-    gtkrc.remove();
-    gtkrc.open(QIODevice::WriteOnly | QIODevice::Text);
-    gtkrc.write(gtkrcContents.toUtf8());
+
+    if (Utils::removeUnlessNixStorePath(gtkrc)) {
+      gtkrc.open(QIODevice::WriteOnly | QIODevice::Text);
+      gtkrc.write(gtkrcContents.toUtf8());
+    };
 }
 
 void removeLegacyStrings()
@@ -74,9 +76,10 @@ void removeLegacyStrings()
     static const QRegularExpression userFontStyleRegexp(QStringLiteral("style(.|\n)*{(.|\n)*}\nwidget_class.*\"user-font\""));
     gtkrcContents.remove(userFontStyleRegexp);
 
-    gtkrc.remove();
-    gtkrc.open(QIODevice::WriteOnly | QIODevice::Text);
-    gtkrc.write(gtkrcContents.toUtf8());
+    if (Utils::removeUnlessNixStorePath(gtkrc)) {
+      gtkrc.open(QIODevice::WriteOnly | QIODevice::Text);
+      gtkrc.write(gtkrcContents.toUtf8());
+    };
 }
 
 }
diff --git a/kded/config_editor/utils.cpp b/kded/config_editor/utils.cpp
index 7897bf5..d0a5fdc 100644
--- a/kded/config_editor/utils.cpp
+++ b/kded/config_editor/utils.cpp
@@ -23,4 +23,13 @@ QString readFileContents(QFile &file)
     }
 }
 
+bool removeUnlessNixStorePath(QFile &file)
+{
+  auto storeDir = qEnvironmentVariable("NIX_STORE_DIR", "/nix/store");
+  if (file.symLinkTarget().startsWith(storeDir)) {
+    return false;
+  }
+  return file.remove();
+}
+
 }
diff --git a/kded/config_editor/utils.h b/kded/config_editor/utils.h
index c8b60c7..6d9e521 100644
--- a/kded/config_editor/utils.h
+++ b/kded/config_editor/utils.h
@@ -10,6 +10,7 @@ namespace Utils
 {
 QString configDirPath(int gtkVersion);
 QString readFileContents(QFile &file);
+bool removeUnlessNixStorePath(QFile &file);
 
 [[maybe_unused]] static inline const auto s_gtkVersions = {3, 4};
 }
diff --git a/kded/config_editor/xsettings.cpp b/kded/config_editor/xsettings.cpp
index 1f9fe5b..fa22108 100644
--- a/kded/config_editor/xsettings.cpp
+++ b/kded/config_editor/xsettings.cpp
@@ -97,9 +97,11 @@ void setValue(const QString &paramName, const QVariant &paramValue)
     QFile xSettingsdConfig(xSettingsdConfigPath);
     QString xSettingsdConfigContents = Utils::readFileContents(xSettingsdConfig);
     replaceValueInXSettingsdContents(xSettingsdConfigContents, paramName, paramValue);
-    xSettingsdConfig.remove();
-    xSettingsdConfig.open(QIODevice::WriteOnly | QIODevice::Text);
-    xSettingsdConfig.write(xSettingsdConfigContents.toUtf8());
+
+    if (Utils::removeUnlessNixStorePath(xSettingsdConfig)) {
+      xSettingsdConfig.open(QIODevice::WriteOnly | QIODevice::Text);
+      xSettingsdConfig.write(xSettingsdConfigContents.toUtf8());
+    };
 
     if (s_applyId == 0) {
 #if GLIB_CHECK_VERSION(2, 74, 0)
```

</details>

And used it in an overlay like this:

<details>

```nix
kdePackages = prev.kdePackages.overrideScope (kfinal: kprev: {
  kde-gtk-config = kprev.kde-gtk-config.overrideAttrs (old: {
    patches = (old.patches or []) ++ [ ./kde-gtk-config-respect-store-symlinks.patch ];
  });
});
```

</details>

## Notes
The patch isn't doing the Nix-store check for any of the `colors.css` files, because I just noticed I had missed those. That said, for those files the module actually does check whether it has write permissions _before_ deleting it and writing a new file in its place. Since the store is read-only, that should already do the right thing.

That aside, I will be adding the Nix-store check explicitly anyway when I have some more free time, just to err on the side of caution.

Also note: I haven't tested this with non-standard Nix store locations. It does accept the path given in the `NIX_STORE_DIR` environment variable, but I'm not sure if this covers all possible weird edge-cases. Alternatively, you could make it Nix-agnostic by just following the symlink and checking whether the destinationfile is writable or not. If not, ignore it. Actually, that might be a patch that you could make a convincing case for merging upstream.

---

EDIT: Fixed an oopsie in the patch that caused it to ignore the contents of the environment variable.
